homeassistant/components/fritzbox_callmonitor/sensor.py

```diff
@@ -1,114 +1,137 @@
-        vol.Optional(CONF_PHONEBOOK, default=0): cv.positive_int,
-        vol.Optional(CONF_PREFIXES, default=[]): vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_PHONEBOOK, default=DEFAULT_PHONEBOOK): cv.positive_int,
+        vol.Optional(CONF_PREFIXES): vol.All(cv.ensure_list, [cv.string]),
     }
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up Fritz!Box call monitor sensor platform."""
-    name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
-    # Try to resolve a hostname; if it is already an IP, it will be returned as-is
-    try:
-        host = socket.gethostbyname(host)
-    except OSError:
-        _LOGGER.error("Could not resolve hostname %s", host)
-        return
-    port = config.get(CONF_PORT)
-    username = config.get(CONF_USERNAME)
-    password = config.get(CONF_PASSWORD)
-    phonebook_id = config.get(CONF_PHONEBOOK)
-    prefixes = config.get(CONF_PREFIXES)
-
-    try:
-        phonebook = FritzBoxPhonebook(
-            host=host,
-            port=port,
-            username=username,
-            password=password,
-            phonebook_id=phonebook_id,
-            prefixes=prefixes,
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Import the platform into a config entry."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
         )
-    except:  # noqa: E722 pylint: disable=bare-except
-        phonebook = None
-        _LOGGER.warning("Phonebook with ID %s not found on Fritz!Box", phonebook_id)
-
-    sensor = FritzBoxCallSensor(name=name, phonebook=phonebook)
-
-    add_entities([sensor])
-
-    monitor = FritzBoxCallMonitor(host=host, port=port, sensor=sensor)
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the fritzbox_callmonitor sensor from config_entry."""
+    phonebook = hass.data[DOMAIN][config_entry.entry_id]
+
+    phonebook_id = config_entry.data[CONF_PHONEBOOK]
+    prefixes = config_entry.data[CONF_PREFIXES]
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
+
+    phonebook_info = await hass.async_add_executor_job(
+        partial(phonebook.fph.phonebook_info, phonebook_id)
+    )
+
+    phonebook_name = phonebook_info.get(FRITZ_ATTR_NAME)
+    phonebook_url = phonebook_info.get(FRITZ_ATTR_URL)
+
+    sensor = FritzBoxCallSensor(
+        host=host,
+        phonebook=phonebook,
+        phonebook_name=phonebook_name,
+        phonebook_url=phonebook_url,
+        phonebook_id=phonebook_id,
+        prefixes=prefixes,
+    )
+
+    monitor = FritzBoxCallMonitor(
+        host=host,
+        port=port,
+        sensor=sensor,
+    )
     monitor.connect()
```

Comment:
You're doing I/O in the event loop.

Reply:
This class should be instantiated inside the entities `async_added_to_hass` lifecycle method . That way if the entity is disabled, we won't set up a connection.

Reply:
Good catch! Will look into it and try to implement your suggestion.

Reply:
Implemented in https://github.com/home-assistant/core/pull/40736/commits/8a0bed4ad86fd358f21f76aee539968f7eddbdb9